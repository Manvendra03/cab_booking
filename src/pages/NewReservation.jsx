/* eslint-disable object-curly-newline */
/* eslint-disable react/jsx-props-no-spreading */
import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import { useLocation, useNavigate } from 'react-router-dom';
import { Circles } from 'react-loader-spinner';
import { useDispatch, useSelector } from 'react-redux';
import { createReservation } from '../redux/slices/reservationSlice';

const NewReservation = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [loading, setLoading] = useState(false);
  const { state } = useLocation();
  const { user } = useSelector((state) => state.user);
  const { cabs } = useSelector((state) => state.fetchCab);
  const [selectedCab, setSelectedCab] = useState('');

  const {
    register,
    handleSubmit,
    formState: { errors },
    reset,
  } = useForm();

  const handleFormSubmit = async ({ date, cab, city }) => {
    setLoading(true);
    const obj = { user_id: user.id, reserve_date: date, cab_id: cab, city };
    const result = await dispatch(createReservation(obj));
    if (result.meta.requestStatus === 'fulfilled') {
      setSelectedCab('');
      reset();
      navigate('/reservations');
    }
    setLoading(false);
  };

  return (
    <main className="flex w-full h-screen px-4 bg-reserve-bg bg-reserve-color bg-blend-multiply text-white bg-center bg-cover overflow-auto items-center gap-12 justify-center pt-10">
      <section className="bg-lime-50 px-4 w-full md:w-auto py-16 pb-4 text-gray-700 rounded-md">
        <h1 className="flex tracking-widest text-lime-800 w-full  justify-center text-xl md:text-3xl lg:text-4xl mb-8 font-bold">
          Book a Ride with Carbooky!
        </h1>
        <form
          onSubmit={handleSubmit(handleFormSubmit)}
          className="w-full mb-12 text-lime-800"
        >
          <div className="flex flex-col">
            <label htmlFor="city">
              City
              <input
                type="text"
                name="city"
                id="city"
                {...register('city', {
                  required: 'Please enter your city',
                })}
              />
            </label>
            {errors.city && (
              <span className="text-red-500 text-base">
                {errors.city.message}
              </span>
            )}
          </div>
          <div className="flex flex-col">
            <label htmlFor="date">
              Date
              <input
                type="date"
                name="date"
                id="date"
                {...register('date', {
                  required: 'Please select a date',
                })}
              />
            </label>
            {errors.date && (
              <span className="text-red-500 text-base">
                {errors.date.message}
              </span>
            )}
          </div>
          <div className="flex flex-col">
            <label htmlFor="Cab">
              Cab Model
              <select
                name="cab"
                id="cab"
                {...register('cab', {
                  required: 'Please choose a cab',
                })}
                value={state?.model || selectedCab}
                onChange={(e) => setSelectedCab(e.target.value)}
              >
                <option value="" disabled>
                  Select a cab
                </option>
                {cabs.map((cab) => (
                  <option key={cab.id} value={cab.id}>
                    {cab.model}
                  </option>
                ))}
              </select>
            </label>
            {errors.cab && (
              <span className="text-red-500 text-base">
                {errors.cab.message}
              </span>
            )}
          </div>
          <div className="flex flex-col">
            <label htmlFor="Cab">
              Username
              <input
                type="text"
                name="name"
                id="name"
                defaultValue={user?.name}
                {...register('name', {
                  required: 'Please choose a name',
                })}
              />
            </label>
            {errors.name && (
              <span className="text-red-500 text-base">
                {errors.name.message}
              </span>
            )}
          </div>

          <div>
            {loading ? (
              <div className="flex justify-center items-center mt-8">
                <Circles
                  height="40"
                  width="40"
                  color="rgba(101, 163, 13, 1)"
                  ariaLabel="circles-loading"
                  wrapperStyle={{}}
                  wrapperClass=""
                  visible={loading}
                />
              </div>
            ) : (
              <button
                type="submit"
                className="bg-lime-600 hover:bg-lime-700 transition-all duration-300 ease-linear w-full mt-8 flex gap-2 items-center text-gray-100 rounded-md justify-center text-2xl px-8 py-3"
              >
                Reserve Now
              </button>
            )}
          </div>
        </form>
      </section>
    </main>
  );
};

export default NewReservation;
