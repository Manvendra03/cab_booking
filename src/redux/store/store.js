import { configureStore } from '@reduxjs/toolkit';
import userReducer from '../slices/userSlice';
import fetchCabReducer from '../slices/cabSlice';
import reservationReducer from '../slices/reservationSlice';
import menuReducer from '../slices/menuSlice';

const store = configureStore({
  reducer: {
    fetchCab: fetchCabReducer,
    user: userReducer,
    reservation: reservationReducer,
    menu: menuReducer,
  },
});

export default store;
