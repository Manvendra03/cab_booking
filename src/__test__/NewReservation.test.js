import { render, screen, waitFor } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../redux/store/store';
import NewReservation from '../pages/NewReservation';

describe('New Reservation', () => {
  it('renders', () => {
    const tree = TestRenderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <NewReservation />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders New Reservation', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <NewReservation />
        </BrowserRouter>
      </Provider>,
    );

    await waitFor(() => {
      const header = screen.getByText(/Book a Ride with Carbooky!/i);
      expect(header).toBeInTheDocument();
    });
  });
  it('renders CabIndex and checks for the subheader', async () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <NewReservation />
        </BrowserRouter>
      </Provider>,
    );

    await waitFor(() => {
      const subHead = screen.getByText(/Cab Model/i);
      expect(subHead).toBeInTheDocument();
    });
  });
});
