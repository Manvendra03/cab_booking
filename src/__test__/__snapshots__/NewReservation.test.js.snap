// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Reservation renders 1`] = `
<main
  className="flex w-full h-screen px-4 bg-reserve-bg bg-reserve-color bg-blend-multiply text-white bg-center bg-cover overflow-auto items-center gap-12 justify-center pt-10"
>
  <section
    className="bg-lime-50 px-4 w-full md:w-auto py-16 pb-4 text-gray-700 rounded-md"
  >
    <h1
      className="flex tracking-widest text-lime-800 w-full  justify-center text-xl md:text-3xl lg:text-4xl mb-8 font-bold"
    >
      Book a Ride with Carbooky!
    </h1>
    <form
      className="w-full mb-12 text-lime-800"
      onSubmit={[Function]}
    >
      <div
        className="flex flex-col"
      >
        <label
          htmlFor="city"
        >
          City
          <input
            id="city"
            name="city"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
          />
        </label>
      </div>
      <div
        className="flex flex-col"
      >
        <label
          htmlFor="date"
        >
          Date
          <input
            id="date"
            name="date"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
          />
        </label>
      </div>
      <div
        className="flex flex-col"
      >
        <label
          htmlFor="Cab"
        >
          Cab Model
          <select
            id="cab"
            name="cab"
            onBlur={[Function]}
            onChange={[Function]}
            value=""
          >
            <option
              disabled={true}
              value=""
            >
              Select a cab
            </option>
          </select>
        </label>
      </div>
      <div
        className="flex flex-col"
      >
        <label
          htmlFor="Cab"
        >
          Username
          <input
            id="name"
            name="name"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
          />
        </label>
      </div>
      <div>
        <button
          className="bg-lime-600 hover:bg-lime-700 transition-all duration-300 ease-linear w-full mt-8 flex gap-2 items-center text-gray-100 rounded-md justify-center text-2xl px-8 py-3"
          type="submit"
        >
          Reserve Now
        </button>
      </div>
    </form>
  </section>
</main>
`;
